README.md

```diff
@@ -100,7 +100,9 @@ We decided to use a [continuation-passing style (CPS)] internally and expose nea
 
 ### DI
 
-In Rust, it is possible to have a type-safe DI container instead of `DependencyMap` that panics at run-time. However, this would require complex type-level manipulations (like those in the [frunk] library). [@p0lunin] and I ([@Hirrolot]) decided not to trade comprehensible error messages for compile-time safety, since we had a plenty of experience that the uninitiated users simply cannot understand what is wrong with their code, owing to the utterly inadequate diagnostic messages from rustc.
+In Rust, it is possible to express type-safe DI that checks all types statically. However, this would require complex type-level manipulations (like those in the [frunk] library). [@p0lunin] and I ([@Hirrolot]) decided not to trade comprehensible error messages for compile-time safety, since we had a plenty of experience that the uninitiated users simply cannot understand what is wrong with their code, owing to the utterly inadequate diagnostic messages from rustc.
+
+The approach taken by `dptree` is to implement run-time type checking instead, via `dptree::type_check`, to make sure that all required types are provided _before_ execution. If `dptree::type_check` is _not_ called, type checking will be delayed until execution (this is not recommended). This approach works much like static type checking when you build your whole dispatch tree at a program startup; the downside is that the panic message raised by `dptree::type_check` does not show the exact location to be fixed, only a final type mismatch message (the types in question are shown though!).
```

Comment:
By the way, I believe that the sentence saying that the panic message doesn't show "the exact location to be fixed" is not appropriate here. Even if the types were checked statically, we wouldn't be able to spot the location of an "error". Imagine that handler `F` is chained to handler `G`; `G` requires type `A` to be present, but `F` does not provide this type to `G`. Is this a type error or not? We can't decide because it might be the case that `F` should have provided this type or the whole resulting handler is meant to require type `A` to be present. This is why we don't type check inside `chain` and `branch`.

Reply:
We can still save the information of where the requirement comes from via [`Location`](https://doc.rust-lang.org/std/panic/struct.Location.html) this of course requires a more complicated structure, but IMO worth it.

Reply:
How are you suggesting to use `Location`? We don't even panic inside `.chain` or `.branch` (except the case where the second branch is `dptree::entry()`).

Reply:
I propose that when we save "obligations" ("input to this handler must contain type `A`"), we also save where they are coming from (akin to rustc's [`ObligationCause`](https://github.com/rust-lang/rust/blob/91b61a4ad618c1abc2af43a58695de185ef1e513/compiler/rustc_middle/src/traits/mod.rs#L96))

Reply:
I've just pushed a commit addressing this issue. Thanks for showing a useful direction.